services/meta/data.go

```diff
@@ -844,6 +835,91 @@ func (di *DatabaseInfo) unmarshal(pb *internal.DatabaseInfo) {
 	}
 }
 
+// RetentionPolicySpec represents the specification for a new retention policy.
+type RetentionPolicySpec struct {
+	Name               string
+	ReplicaN           *int
+	Duration           *time.Duration
+	ShardGroupDuration time.Duration
+}
+
+// NewRetentionPolicyInfo creates a new retention policy info from the specification.
+func (s *RetentionPolicySpec) NewRetentionPolicyInfo() *RetentionPolicyInfo {
+	return DefaultRetentionPolicyInfo().Apply(s)
+}
+
+// Matches checks if this retention policy specification matches
+// an existing retention policy.
+func (s *RetentionPolicySpec) Matches(rpi *RetentionPolicyInfo) bool {
+	if s.Name != "" && s.Name != rpi.Name {
+		return false
+	} else if s.Duration != nil && *s.Duration != rpi.Duration {
+		return false
+	} else if s.ReplicaN != nil && *s.ReplicaN != rpi.ReplicaN {
+		return false
+	}
+
+	// Normalise ShardDuration before comparing to any existing retention policies.
+	// Normalize with the retention policy info's duration instead of the spec
+	// since they should be the same and we're performing a comparison.
+	sgDuration := normalisedShardDuration(s.ShardGroupDuration, rpi.Duration)
+	if sgDuration != rpi.ShardGroupDuration {
+		return false
+	}
+	return true
+}
+
+// marshal serializes to a protobuf representation.
+func (s *RetentionPolicySpec) marshal() *internal.RetentionPolicySpec {
+	pb := &internal.RetentionPolicySpec{}
+	if s.Name != "" {
+		pb.Name = proto.String(s.Name)
+	}
+	if s.Duration != nil {
+		pb.Duration = proto.Int64(int64(*s.Duration))
+	}
+	if s.ShardGroupDuration > 0 {
+		pb.ShardGroupDuration = proto.Int64(int64(s.ShardGroupDuration))
+	}
+	if s.ReplicaN != nil {
+		pb.ReplicaN = proto.Uint32(uint32(*s.ReplicaN))
+	}
+	return pb
+}
+
+// unmarshal deserializes from a protobuf representation.
+func (s *RetentionPolicySpec) unmarshal(pb *internal.RetentionPolicySpec) {
+	if pb.Name != nil {
+		s.Name = pb.GetName()
+	}
+	if pb.Duration != nil {
+		duration := time.Duration(pb.GetDuration())
+		s.Duration = &duration
+	}
+	if pb.ShardGroupDuration != nil {
+		s.ShardGroupDuration = time.Duration(pb.GetShardGroupDuration())
+	}
+	if pb.ReplicaN != nil {
+		replicaN := int(pb.GetReplicaN())
+		s.ReplicaN = &replicaN
+	}
+}
+
+// MarshalBinary encodes RetentionPolicySpec to a binary format.
+func (s *RetentionPolicySpec) MarshalBinary() ([]byte, error) {
+	return proto.Marshal(s.marshal())
+}
+
+// UnmarshalBinary decodes RetentionPolicySpec from a binary format.
+func (s *RetentionPolicySpec) UnmarshalBinary(data []byte) error {
+	var pb internal.RetentionPolicySpec
+	if err := proto.Unmarshal(data, &pb); err != nil {
+		return err
+	}
+	s.unmarshal(&pb)
+	return nil
+}
+
 // RetentionPolicyInfo represents metadata about a retention policy.
 type RetentionPolicyInfo struct {
```

Comment:
`RetentionPolicyInfo` or `RetentionPolicySpec`? For `Info`, it's because that's what it's named everywhere and the struct encapsulates the info for a retention policy. There's also a `DatabaseInfo` struct too. I'm not sure why it was originally done that way, but that's what's in the code in a number of different places.

For the spec, I didn't want to confuse anybody by calling it a `RetentionPolicy` since it isn't. It's a specification for how to make a retention policy.